### PR TITLE
Adopt tiered code metrics enforcement

### LINKED ISSUE #194: Evaluate and adopt legacy CA/code-metrics rules with tiered enforcement
## Background

The repository already uses:

- `EnableNETAnalyzers=true`
- `AnalysisMode=All`
- `EnforceCodeStyleInBuild=true`
- `CodeAnalysisTreatWarningsAsErrors=true`

This means the normal .NET analyzer set is already enforced strictly. However, several legacy / code-metrics CA rules are not automatically enabled by `AnalysisMode=All`, so an isolated inventory scan was run against a temporary repository copy.

The scan did not modify the original repository and the build completed successfully.

## Inventory result

| Rule | Build-output count | Approx. unique | Initial assessment |
| --- | ---: | ---: | --- |
| CA1005 | 0 | 0 | Suitable for blocking gate |
| CA1017 | 0 | 0 | Suitable for blocking gate |
| CA1021 | 0 | 0 | Suitable for blocking gate |
| CA1045 | 2 | 1 | Actionable |
| CA1060 | 2 | 1 | Actionable |
| CA1501 | 104 | ~52 | Mostly framework-driven noise |
| CA1502 | 0 | 0 | Suitable for blocking gate |
| CA1505 | 0 | 0 | Suitable for blocking gate |
| CA1506 | 26 | ~13 | Useful architecture signal |
| CA1509 | 0 | 0 | Suitable for blocking gate |

Note: `dotnet build` repeats diagnostics in compile output / summary, so the raw counts above are not unique-defect counts.

### CA1045

Known location:

`DownKyi.Core/Utils/ListHelper.cs`

`InsertUnique<T>(..., ref T currentSelection)`

Trace all call sites and determine whether the `ref` parameter has any real observable behavioral requirement. If it is historical API residue, remove the unnecessary by-ref contract while preserving collection ordering / selection behavior and adding or adjusting regression coverage.

### CA1060

Known location:

`tests/DownKyi.Tests/Aria2TlsTestRuntime.cs`

Move P/Invoke declarations into an appropriate native interop holder such as `NativeMethods`, without changing native signatures, calling conventions, platform behavior, or TLS integration coverage.

### CA1506

Known production hotspots include:

- `FfmpegConcatRuntime.ConcatAsync`
- `SettingsManager`
- `Aria2TransferBackend.TransferAsync`
- `BuiltinTransferBackend.TransferAsync`
- `DownloadRuntimeFactory.Create`

Examples observed during the scan:

- `FfmpegConcatRuntime.ConcatAsync`: 44 coupled types, threshold 41
- `DownloadRuntimeFactory.Create`: 50 coupled types, threshold 41
- `SettingsManager`: 104 coupled types, threshold 96

These findings must not be treated automatically as defects. Classify them as one of:

- real architecture hotspot
- legitimate composition root
- integration-test coupling
- framework-driven coupling
- needs manual review

CA1506 should therefore be used as an advisory architecture audit rather than as a blocking build rule.

### CA1501

Most CA1501 findings are caused by normal Avalonia framework inheritance, for example:

`MainWindow -> Window -> WindowBase -> TopLevel -> ContentControl -> TemplatedControl -> Control -> ... -> AvaloniaObject`

and many normal `SomeView : UserControl` types.

This is framework inheritance, not a DownKyi-designed 10+ level application hierarchy.

Do not:

- rewrite normal Avalonia UI inheritance to satisfy CA1501
- create dozens of low-value maintenance tasks for framework-driven findings
- add one suppression per Avalonia view simply to force a zero count

Only review CA1501 when it points to a deep inheritance hierarchy controlled by DownKyi itself.

### Tier 1 — Blocking gate

After cleaning any existing findings, explicitly set these rules to `error`:

- CA1005
- CA1017
- CA1021
- CA1045
- CA1060
- CA1502
- CA1505
- CA1509

CA1045 and CA1060 should be fixed first, then made blocking.

Do not use project-wide `NoWarn`, global suppression, or `severity=none` to manufacture a clean baseline.

### Tier 2 — Non-blocking advisory

Keep CA1506 as an architecture advisory.

Because the normal build uses warnings-as-errors, do not simply set CA1506 to warning in the main build pipeline. Instead, add an isolated audit path, for example:

`script/audit-code-metrics.ps1`

The audit should:

1. Enable CA1506 explicitly.
2. Use `TreatWarningsAsErrors=false`.
3. Use `CodeAnalysisTreatWarningsAsErrors=false`.
4. Collect and deduplicate CA1506 diagnostics.
5. Separate production findings from test findings.
6. Produce a human-readable summary.
7. Preferably also produce a machine-readable JSON report.
8. Exit 0 when CA1506 findings are present.
9. Exit non-zero only when the audit infrastructure itself fails or no report can be produced.

Production `ArchitectureHotspot` findings may enter the maintenance backlog after manual review.

In particular, review `DownloadRuntimeFactory.Create` to determine whether the coupling represents:

- a legitimate composition root; or
- excessive orchestration / construction responsibility.

Do not split a legitimate composition root into artificial thin wrappers simply to satisfy a metric.

Apply the same reasoning to `SettingsManager`, including whether partial-class/settings aggregation inflates the metric without implying a real maintenance defect.

### Tier 3 — Observation only

Do not add CA1501 to the normal blocking CI.

It may remain in a manual metrics audit or be limited to DownKyi-controlled inheritance structures.

## Acceptance criteria

**Status: Completed.** All acceptance criteria have been satisfied on `main`.

- [x] Existing unique CA1045 finding is analyzed and fixed, or retained with a concrete contract-based justification
- [x] Existing unique CA1060 finding is fixed
- [x] CA1005 / CA1017 / CA1021 / CA1045 / CA1060 / CA1502 / CA1505 / CA1509 are part of the blocking analyzer policy
- [x] No project-wide `NoWarn`
- [x] No global suppression used to hide the baseline
- [x] CA1506 has an independent non-blocking audit path
- [x] CA1506 reporting separates production / test findings
- [x] CA1506 findings are classified by actual responsibility semantics rather than metric magnitude alone:
  - `operation-owner`: complexity belongs to one complete operation whose invariants must remain centralized, such as concat or a single transfer.
  - `settings-aggregation`: the metric is inflated by partial-class or settings aggregation rather than a standalone architecture defect.
  - `composition-root`: complexity belongs to legitimate dependency construction and runtime composition.
  - `integration-test`: complexity belongs to integration coverage that must coordinate multiple components.
  - `needs-review`: there is not yet enough evidence to classify the finding as reasonable or actionable.
  
  A reviewed finding in one of the reasonable categories must not be split merely to reduce CA1506. Refactor only when there is independent architecture or maintainability evidence.
- [x] CA1501 does not require refactoring normal Avalonia framework inheritance
- [x] Release build remains warnings-as-errors
- [x] Full solution tests pass
- [x] `dotnet format --verify-no-changes` passes
- [x] `git diff --check` passes

## Non-goals

This issue does not require:

- large-scale UI framework refactoring just to make legacy CA counts reach zero
- enabling every legacy rule unconditionally as an error
- dismantling normal Avalonia inheritance for CA1501
- splitting legitimate composition roots into artificial wrappers for CA1506
- changing product behavior, download behavior, settings formats, or public runtime contracts

### PR DESCRIPTION
**Pull request by crazysmile-PhD**
## Summary

- fix the CA1045 and CA1060 findings while preserving ViewModel selection behavior and Windows native interop behavior
- make CA1005, CA1017, CA1021, CA1045, CA1060, CA1502, CA1505, and CA1509 blocking analyzer rules
- add an isolated non-blocking CA1506 audit with deterministic production/test classification, Markdown and JSON reports, and fail-closed infrastructure behavior
- document and test the extremely thin PowerShell boundary; all report logic remains in C#

## Validation

- strict Release analyzer build: 0 warnings, 0 errors
- full Windows solution tests: 985 passed, 0 failed, 1 skipped (packaged aria2 binary not present locally)
- CA1506 audit: 13 advisory findings (5 production, 8 test), exit 0
- dotnet format --verify-no-changes
- module-boundary audit
- actionlint v1.7.12
- vulnerable/deprecated package audits
- Gitleaks 8.30.1: 0 findings
- git diff --check

Closes #194

### CODE REVIEW

**Comment by crazysmile-PhD** [comment]
@codex review

File: tools/DownKyi.CodeMetricsAudit/Ca1506ReportWriter.cs [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    public static void Write(string outputDirectory, Ca1506Report report)
+    {
+        ArgumentException.ThrowIfNullOrWhiteSpace(outputDirectory);
+        ArgumentNullException.ThrowIfNull(report);
+        Directory.CreateDirectory(outputDirectory);
+
+        var jsonPath = Path.Combine(outputDirectory, "ca1506-report.json");
+        var markdownPath = Path.Combine(outputDirectory, "ca1506-report.md");
+        File.WriteAllText(jsonPath, JsonSerializer.Serialize(report, JsonOptions) + "\n", new UTF8Encoding(false));
+        File.WriteAllText(markdownPath, BuildMarkdown(report), new UTF8Encoding(false));
```
[the remark this answers is not in the document]
Comment (crazysmile-PhD) [reply tools/DownKyi.CodeMetricsAudit/Ca1506ReportWriter.cs]: Fixed at exact HEAD 9967f92e. Both reports are staged first, existing artifacts are backed up, and publication rolls back on either move failure. The failure-injection regression verifies both previous reports remain intact and no staging files remain; the exact-head CA1506 audit is green.

File: tools/DownKyi.CodeMetricsAudit/Program.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+                $"Production: {report.Summary.Production}; test: {report.Summary.Test}").ConfigureAwait(false);
+            await Console.Out.WriteLineAsync(
+                $"Markdown: {Path.Combine(options.OutputDirectory, "ca1506-report.md")}").ConfigureAwait(false);
+            await Console.Out.WriteLineAsync(
+                $"JSON: {Path.Combine(options.OutputDirectory, "ca1506-report.json")}").ConfigureAwait(false);
+            return 0;
+        }
+        catch (Exception exception) when (exception is ArgumentException
+            or IOException
+            or InvalidDataException
```
[the remark this answers is not in the document]
Comment (crazysmile-PhD) [reply tools/DownKyi.CodeMetricsAudit/Program.cs:41]: Fixed at exact HEAD 9967f92e. Program.Main now catches JsonException and returns 1 with a bounded fixed message. The regression uses syntactically invalid {broken input and verifies no exception name, stack trace, or repository path is emitted.

File: tools/DownKyi.CodeMetricsAudit/Ca1506ReportGenerator.cs [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+        {
+            return new ClassificationDecision(
+                "test",
+                "test integration",
+                "The finding is in integration or behavioral test composition rather than production runtime code.");
+        }
+
+        if (classifications.TryGetValue(file, out var classification))
+        {
+            return new ClassificationDecision("production", classification.Classification, classification.Rationale);
```
[the remark this answers is not in the document]
Comment (crazysmile-PhD) [reply tools/DownKyi.CodeMetricsAudit/Ca1506ReportGenerator.cs]: Fixed at exact HEAD 9967f92e. Schema v2 classification now uses repository-relative file plus source-region identity (location:line:column), so each diagnostic is reviewed independently and an unmatched production finding remains needs manual review.

File: tools/DownKyi.CodeMetricsAudit/GitStateReader.cs [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+{
+    public static async Task<GitState> ReadAsync(string repositoryRoot)
+    {
+        ArgumentException.ThrowIfNullOrWhiteSpace(repositoryRoot);
+        var commit = await RunGitAsync(repositoryRoot, "rev-parse", "HEAD").ConfigureAwait(false);
+        var status = await RunGitAsync(
+                repositoryRoot,
+                "status",
+                "--porcelain",
+                "--untracked-files=no")
```
[the remark this answers is not in the document]
Comment (crazysmile-PhD) [reply tools/DownKyi.CodeMetricsAudit/GitStateReader.cs]: Fixed at exact HEAD 9967f92e. Provenance now uses git status --porcelain --untracked-files=all. The regression creates an untracked SDK-style .cs compile input and verifies dirtyWorktree is true.

File: tools/DownKyi.CodeMetricsAudit/Program.cs [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+            Ca1506ReportWriter.Write(options.OutputDirectory, report);
+
+            await output.WriteLineAsync(
+                $"CA1506 audit completed with {report.Findings.Count} advisory finding(s).").ConfigureAwait(false);
+            await output.WriteLineAsync(
+                $"Production: {report.Summary.Production}; test: {report.Summary.Test}").ConfigureAwait(false);
+            await output.WriteLineAsync(
+                $"Markdown: {Path.Combine(options.OutputDirectory, "ca1506-report.md")}").ConfigureAwait(false);
+            await output.WriteLineAsync(
+                $"JSON: {Path.Combine(options.OutputDirectory, "ca1506-report.json")}").ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (crazysmile-PhD) [reply tools/DownKyi.CodeMetricsAudit/Program.cs]: Fixed at exact HEAD 9967f92e. The success message now emits only fixed artifact filenames, and the regression verifies it contains neither the repository root nor the user-home prefix while both artifacts are written to the requested output directory.

File: tools/DownKyi.CodeMetricsAudit/Ca1506ReportGenerator.cs [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+
+    private static string ReadDiagnosticIdentity(string message, int line, int column)
+    {
+        var openingQuote = message.IndexOf('\'', StringComparison.Ordinal);
+        if (openingQuote >= 0)
+        {
+            var closingQuote = message.IndexOf('\'', openingQuote + 1);
+            if (closingQuote > openingQuote + 1)
+            {
+                return message[(openingQuote + 1)..closingQuote];
```
[the remark this answers is not in the document]
Comment (crazysmile-PhD) [reply tools/DownKyi.CodeMetricsAudit/Ca1506ReportGenerator.cs]: Fixed at exact HEAD 9967f92e. The key is repository-relative file plus deterministic source region, not an unqualified symbol. Regression coverage includes same-named methods on different types, an overload at a different location, rationale preservation for the reviewed finding, manual-review fallback for new findings, and order independence.

File: tests/DownKyi.Windows.Tests/TargetedResourceForensicsWindowsTests.cs [hunk 7]
```diff
@@ -419,6 +409,32 @@ private static Process StartDirectoryOwner(string workingDirectory)
...
+            UseShellExecute = false,
+            RedirectStandardOutput = true,
+            RedirectStandardError = true,
+            CreateNoWindow = true,
+        };
+        startInfo.ArgumentList.Add("exec");
+        startInfo.ArgumentList.Add("--runtimeconfig");
+        startInfo.ArgumentList.Add(runtimeConfig);
+        startInfo.ArgumentList.Add(fixtureAssembly);
+        startInfo.ArgumentList.Add("fixture-directory-lock");
```
[the remark this answers is not in the document]
Comment (crazysmile-PhD) [reply tests/DownKyi.Windows.Tests/TargetedResourceForensicsWindowsTests.cs:432]: Fixed at exact HEAD 9967f92e by matching the assertion to fixture-directory-lock. Both the exact-head hosted Windows targeted-resource-forensics step and the subsequent full Test step passed.

**Comment by crazysmile-PhD** [comment]
@codex review

Please review the current exact HEAD `9967f92e6907288f86780fbbb02e2f9c143e38b9`.

Previous review findings have been addressed and all existing review threads are resolved. Strict PR CI and CodeQL are green.

Please review the current HEAD as the final pre-merge review. Do not rely on the earlier `5869a7f` review.

File: tools/DownKyi.CodeMetricsAudit/GitStateReader.cs [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+{
+    public static async Task<GitState> ReadAsync(string repositoryRoot)
+    {
+        ArgumentException.ThrowIfNullOrWhiteSpace(repositoryRoot);
+        var commit = await RunGitAsync(repositoryRoot, "rev-parse", "HEAD").ConfigureAwait(false);
+        var status = await RunGitAsync(
+                repositoryRoot,
+                "status",
+                "--porcelain",
+                "--untracked-files=all")
```
[the remark this answers is not in the document]
Comment (crazysmile-PhD) [reply tools/DownKyi.CodeMetricsAudit/GitStateReader.cs:17]: Fixed at exact HEAD 029d9997. GitStateReader now separately enumerates ignored C# files with git ls-files --others --ignored --exclude-standard, excluding only standard bin/obj build-output trees. The regression uses the repository-relevant DownKyi.Core/Binary/* ignore rule and proves an ignored CompileInput.cs marks dirtyWorktree=true. The exact-head CA1506 audit job passed.

File: tools/DownKyi.CodeMetricsAudit/Program.cs [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+            await error.WriteLineAsync("CA1506 audit failed: malformed JSON input.").ConfigureAwait(false);
+            return 1;
+        }
+        catch (Exception exception) when (exception is ArgumentException
+            or IOException
+            or InvalidDataException
+            or InvalidOperationException
+            or UnauthorizedAccessException)
+        {
+            await error.WriteLineAsync($"CA1506 audit failed: {exception.Message}").ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (crazysmile-PhD) [reply tools/DownKyi.CodeMetricsAudit/Program.cs]: Fixed at exact HEAD 029d9997. Controlled non-JSON failures now emit fixed categories instead of exception.Message. Regressions cover a missing absolute input path and a path-bearing output IOException, require nonzero exit, and verify neither repository root nor user-home prefix is disclosed. Architecture tests and the exact-head CA1506 audit job passed.

**Comment by crazysmile-PhD** [comment]
@codex review

File: tools/DownKyi.CodeMetricsAudit/Ca1506ReportWriter.cs [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+                    backupJsonPath,
+                    backupMarkdownPath,
+                    publishedJson,
+                    publishedMarkdown,
+                    exception);
+                throw;
+            }
+
+            File.Delete(backupJsonPath);
+            File.Delete(backupMarkdownPath);
```
[the remark this answers is not in the document]
Comment (crazysmile-PhD) [reply tools/DownKyi.CodeMetricsAudit/Ca1506ReportWriter.cs]: Acknowledged. The current JSON/Markdown pair is already staged and published as one rollback-capable unit; this finding is limited to post-publication cleanup of the previous-report backups. Because deleting two backups is not atomic, a safe fix for second-delete failure requires durable backup ownership/recovery semantics rather than a catch, sleep, or retry. That transaction expansion is disproportionate to #194/#219, so it is tracked in #224 with the residual risk and deterministic acceptance criteria. This thread remains open because the cleanup edge case is not fixed in this PR.

File: tools/DownKyi.CodeMetricsAudit/Ca1506ReportGenerator.cs [hunk 11]
```diff
@@ -0,0 +1,227 @@
...
+            findings.Count(finding => finding.Scope == "test"),
+            classificationCounts);
+        return new Ca1506Report(1, "CA1506", gitState.Commit, gitState.DirtyWorktree, summary, findings);
+    }
+
+    private static Dictionary<string, ProductionClassification> ReadClassifications(string path)
+    {
+        using var document = JsonDocument.Parse(File.ReadAllText(path));
+        var root = document.RootElement;
+        if (!root.TryGetProperty("schemaVersion", out var schemaVersion) || schemaVersion.GetInt32() != 2)
```
[the remark this answers is not in the document]
Comment (crazysmile-PhD) [reply tools/DownKyi.CodeMetricsAudit/Ca1506ReportGenerator.cs]: Fixed at exact HEAD ce97d4d0. All production audit Int32 JSON conversions were audited: schemaVersion, SARIF startLine, and SARIF startColumn now require Number plus TryGetInt32 and fail as controlled invalid audit input. Program also bounds FormatException at the audit boundary. Regressions cover schemaVersion=1e100, non-integral startLine=1.5, and out-of-range startColumn=1e100; each returns 1 with the fixed invalid-input message and no repository/home path. Focused tests passed 15/15, strict build passed with zero warnings/errors, and the exact-head CA1506 audit job passed.

**Comment by crazysmile-PhD** [comment]
@codex review

File: tools/DownKyi.CodeMetricsAudit/Ca1506ReportWriter.cs (hunk 10 above)
[the remark this answers is not in the document]
Comment (crazysmile-PhD) [reply tools/DownKyi.CodeMetricsAudit/Ca1506ReportWriter.cs]: Fixed by 5bc818a7 and retained at exact HEAD 4db1b04e. The previous JSON/Markdown pair is now stored as one operation-owned .backup.zip with fixed entry names and byte validation before publication. Successful publication performs one cleanup delete; injected cleanup failure returns a controlled non-zero result while preserving the complete new report pair and the complete previous pair in one discoverable bundle, so one-sided backup state is no longer possible. Regression coverage also preserves the existing second-publication-failure rollback contract. Focused CA1506 tests passed 26/26 at the current head, strict Release build passed with 0 warnings/errors, format and diff checks passed.

File: script/audit-code-metrics.ps1 [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+        --output-directory $OutputDirectory
+    if ($LASTEXITCODE -ne 0) {
+        throw "CA1506 report generation failed with exit code $LASTEXITCODE."
+    }
+}
+finally {
+    $env:DOTNET_CLI_UI_LANGUAGE = $previousUiLanguage
+    if ($auditRoot.StartsWith($tempBase, [StringComparison]::OrdinalIgnoreCase) -and
+        [IO.Directory]::Exists($auditRoot)) {
+        [IO.Directory]::Delete($auditRoot, $true)
```
[the remark this answers is not in the document]
Comment (crazysmile-PhD) [reply script/audit-code-metrics.ps1]: Fixed at exact HEAD 4db1b04e. CA1506 temporary data now lives under the fixed system-temporary ownership root downkyi-ca1506 with one strictly named 32-hex operation child. Startup scavenging is limited to matching direct children older than 24 hours; locked stale children remain discoverable without blocking cleanup of other stale children or changing the current audit result. Immediate cleanup still runs in finally and remains fail-closed with a fixed path-free message. Dot-dot, paths outside the ownership root, unrelated names, and reparse-point candidates fail safe. Deterministic lifecycle/security regressions and existing CA1506 policy/report tests passed 26/26; strict Release build passed with 0 warnings/errors, format/diff checks passed, and a real audit smoke completed with 13 advisory findings and left zero owned operation directories.

**Comment by crazysmile-PhD** [comment]
@codex review
